lib/Serializer.js

```diff
@@ -0,0 +1,118 @@
+class Serializer {
+	constructor() {
+		this._serializers = {_default: _defaultSerializer};
+		this._defaultSerializer = "_default";
+	}
+
+	add(name, options) {
+		validateName(name);
+		validateOptions(options);
+		this._serializers[name] = options;
+	}
+
+	setDefault(name) {
+		validateName(name);
+		if (Object.prototype.hasOwnProperty.call(this._serializers, name)) {
+			this._defaultSerializer = name;
+		}
+	}
+
+	remove(name) {
+		validateName(name);
+		if (Object.prototype.hasOwnProperty.call(this._serializers, name)) {
+			delete this._serializers[name];
+		}
+		if (this._defaultSerializer === name) {
+			this._defaultSerializer = "_default";
+		}
+	}
+
+	_serializeMany(documentsArray = [], nameOrOptions) {
+		documentsArray = cleanAndValidateDocumentsArray(documentsArray);
+		return documentsArray.map(doc => doc.serialize(nameOrOptions));
+	}
+
+	_serialize(document, nameOrOptions = this._defaultSerializer) {
+		const inputType = typeof nameOrOptions;
+		let isArray = Array.isArray(nameOrOptions);
+		let options;
+
+		if (inputType === "string") {
+			options = this._serializers[nameOrOptions];
+		} else if (isArray || inputType === "object") {
+			options = nameOrOptions;
+		}
+
+		try {
+			validateOptions(options);
+			isArray = Array.isArray(options);
+			if (isArray) {
+				return includeHandler(document, options, {});
+			}
+
+			let serialized = {};
+			if (options.include) {
+				serialized = includeHandler(document, options.include, serialized);
+			}
+			if (options.exclude) {
+				if (!options.include) {
+					serialized = {...document};
+				}
+				serialized = excludeHandler(document, options.exclude, serialized);
+			}
+			if (options.modify && typeof options.modify === "function") {
+				if (!options.include && !options.exclude) {
+					serialized = {...document};
+				}
+				serialized = options.modify(serialized, document);
+			}
+			return serialized;
+		} catch (error) {
+			// Failing quietly and defaulting to dumping the whole object may not be safest idea, lest we expose sensitive data.
+			return {};
```

Comment:
How do we feel about this? What errors are we expecting here? Would it be better to just throw the error?

Reply:
The only case I can forsee is just sending completely wrong data types. Yea throwing an error is probably acceptable, though maybe a specific type of error so its more easily handleable

Reply:
@ablankenship10 That sounds like a great solution to me. I think it'll confuse users more if it returns an empty object on error.

Reply:
Fixed in 1faea335b85769c2e9e07d14459601c385a0f824